File: backend/src/override_identity.rs
//! Override 规则主键的确定性哈希（SHA-256）。规范见仓库根目录 `docs/override-id.md`。
use crate::state::OverrideRule;
use sha2::Digest;
use sha2::Sha256;

/// Same as traffic path normalization: empty -> `/`, ensure leading `/`.
fn normalize_path_for_id(p: &str) -> String {
    let t = p.trim();
    if t.is_empty() {
        return "/".to_string();
    }
    if t.starts_with('/') {
        t.to_string()
    } else {
        format!("/{t}")
    }
}

/// Sorted concatenation: for each (k, v) sorted by (k ascii_lowercase, v), append
/// `k.to_lowercase() + v` in order.
pub fn sorted_kv_blob(pairs: &[(String, String)]) -> String {
    let mut v: Vec<(&str, &str)> = pairs
        .iter()
        .map(|(a, b)| (a.as_str(), b.as_str()))
        .collect();
    v.sort_by(|(ka, va), (kb, vb)| {
        let la = ka.to_ascii_lowercase();
        let lb = kb.to_ascii_lowercase();
        la.cmp(&lb).then_with(|| (*va).cmp(*vb))
    });
    let mut s = String::new();
    for (k, val) in v {
        s.push_str(&k.to_ascii_lowercase());
        s.push_str(val);
    }
    s
}

/// Canonical string hashed into an override id (see product spec: protocol, host, path,
/// sorted header and query key/value, body).
pub fn identity_material(rule: &OverrideRule) -> String {
    let p = rule.match_protocol.as_deref().unwrap_or("");
    let h = rule.match_host.as_deref().unwrap_or("");
    let path = match &rule.match_path {
        None => String::new(),
        Some(s) if s.trim().is_empty() => String::new(),
        Some(s) => normalize_path_for_id(s),
    };
    let hb = sorted_kv_blob(&rule.match_request_headers);
    let qb = sorted_kv_blob(&rule.match_query);
    let b = rule.match_request_body.as_deref().unwrap_or("");
    format!("{p}{h}{path}{hb}{qb}{b}")
}

pub fn override_id_from_material(material: &str) -> String {
    let mut h = Sha256::new();
    h.update(material.as_bytes());
    let out = h.finalize();
    format!("{out:x}")
}

pub fn override_id_for_rule(rule: &OverrideRule) -> String {
    override_id_from_material(&identity_material(rule))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn id_stable() {
        let r = OverrideRule {
            id: "x".to_string(),
            enabled: true,
            match_protocol: Some("https".to_string()),
            match_host: Some("example.com".to_string()),
            match_path: Some("/a".to_string()),
            match_request_headers: vec![("X-Api".to_string(), "1".to_string())],
            match_query: vec![("B".to_string(), "2".to_string())],
            match_request_body: None,
            status: 200,
            headers: vec![],
            body: String::new(),
            stream_interval_ms: None,
        };
        let id1 = override_id_for_rule(&r);
        let id2 = override_id_for_rule(&r);
        assert_eq!(id1, id2);
        assert_eq!(id1.len(), 64);
    }
}

File: frontend/src/lib/overrideIdentity.ts
/**
 * Override id（SHA-256）与后端 `src/override_identity.rs` 对齐；完整规则见 `docs/override-id.md`。
 */
import type { OverrideFormState, OverrideRule } from '../types'
import { normalizePath } from './dashboardUtils'

export function sortedKvBlob(pairs: [string, string][]): string {
  const sorted = [...pairs].sort((a, b) => {
    const la = a[0].toLowerCase()
    const lb = b[0].toLowerCase()
    if (la !== lb) return la.localeCompare(lb, undefined, { sensitivity: 'base' })
    return a[1].localeCompare(b[1], undefined, { sensitivity: 'base' })
  })
  return sorted.map(([k, v]) => k.toLowerCase() + v).join('')
}

/**
 * Canonical string hashed to the override id (mirrors `override_identity` in Rust):
 * method + protocol + host + path + sorted headers/query + body.
 */
export function identityMaterialFromMatch(args: {
  matchMethod: string
  matchProtocol: string
  matchHost: string
  matchPath: string
  matchRequestHeaders: [string, string][]
  matchQuery: [string, string][]
  matchRequestBody: string
}): string {
  const m = args.matchMethod
  const p = args.matchProtocol
  const h = args.matchHost
  const path =
    args.matchPath.trim() === '' ? '' : normalizePath(args.matchPath)
  const hb = sortedKvBlob(args.matchRequestHeaders)
  const qb = sortedKvBlob(args.matchQuery)
  const b = args.matchRequestBody
  return `${m}${p}${h}${path}${hb}${qb}${b}`
}

export async function sha256Hex(utf8: string): Promise<string> {
  const data = new TextEncoder().encode(utf8)
  const buf = await crypto.subtle.digest('SHA-256', data)
  return Array.from(new Uint8Array(buf), (b) => b.toString(16).padStart(2, '0')).join(
    '',
  )
}

export async function computeOverrideIdFromForm(args: {
  matchMethod: string
  matchProtocol: string
  matchHost: string
  matchPath: string
  matchRequestHeaders: [string, string][]
  matchQuery: [string, string][]
  matchRequestBody: string
}): Promise<string> {
  const m = identityMaterialFromMatch(args)
  return sha256Hex(m)
}

function cleanKv(pairs: [string, string][]): [string, string][] {
  return pairs.filter(([a, b]) => a.trim() !== '' || b.trim() !== '')
}

export async function computeOverrideIdFromFormState(
  f: OverrideFormState,
): Promise<string> {
  const matchRequestHeaders = cleanKv(f.matchRequestHeaders)
  const matchQuery = cleanKv(f.matchQuery)
  return computeOverrideIdFromForm({
    matchMethod: f.matchMethod,
    matchProtocol: f.matchProtocol,
    matchHost: f.matchHost,
    matchPath: f.matchPath,
    matchRequestHeaders,
    matchQuery,
    matchRequestBody: f.matchRequestBody,
  })
}

/** JSON body for PUT /api/overrides/:id (shape matches Rust `UpsertOverrideBody`). */
export function apiPayloadFromRule(
  override: OverrideRule,
  body: string,
  enabled: boolean,
) {
  return {
    enabled,
    matchMethod: override.matchMethod?.trim() ? override.matchMethod : null,
    matchProtocol: override.matchProtocol ?? null,
    matchHost: override.matchHost ?? null,
    matchPath: override.matchPath ?? null,
    matchRequestHeaders: override.matchRequestHeaders ?? [],
    matchQuery: override.matchQuery ?? [],
    matchRequestBody: override.matchRequestBody?.trim()
      ? override.matchRequestBody
      : null,
    mapRemoteProtocol: override.mapRemoteProtocol?.trim()
      ? override.mapRemoteProtocol
      : null,
    mapRemoteHost: override.mapRemoteHost?.trim() ? override.mapRemoteHost : null,
    mapRemotePath: override.mapRemotePath?.trim() ? override.mapRemotePath : null,
    status: override.status,
    headers: override.headers,
    body,
    streamIntervalMs: override.streamIntervalMs ?? null,
  }
}
